### DISCUSSION TITLE
Is there a way to get SSR prefetch data as JSON when route is changed?

### DISCUSSION THREAD

**Discussion by a-salimkhanov** [quasarframework/quasar#9477] [CLI - SSR mode]
I get data via `axios` in `prefetch`.
Everything is fine when user loads the whole page. But when `router-link` is clicked, the request to API takes place on client side.

Is there an alternative to Next JS `getStaticProps` that downloads JSON file with needed data of the next page when the router is changed?

And it would be great if this JSON file is cached on server. So that a request is not repeated if the page or JSON file has been generated before

**Comment by hawkeye64** [comment]
@a-salimkhanov Can't you make use of vue-router navigation guards?
https://router.vuejs.org/guide/advanced/navigation-guards.html#in-component-guards

**Reply by a-salimkhanov** [reply]
@hawkeye64 I read about navigation guards and I think that's not exactly what I need (or didn't I get the use of it?)

What I need is realized in Nuxt, I guess ( [https://nuxtjs.org/docs/2.x/features/data-fetching/](url) ) It seems I'll have to convert my project to Nuxt.

I need API requests to be processed only on server-side (from Frontend server to API server).
But when a user goes to another inner link of the website (clicking on `router-link`), the API request (located in prefetch section) is made from client browser to API server directly.

And that exposes information about the API location, credentials and so on.

**Reply by ontwikkelfabriek** [reply]
@a-salimkhanov 
> I need API requests to be processed only on server-side (from Frontend server to API server).

In Qv2 you can do that. See this comment ( some other comments in the discussion are interesting too)
https://github.com/quasarframework/quasar/discussions/9298#discussioncomment-748052

**Reply by a-salimkhanov** [reply]
Zeus bless you, @ontwikkelfabriek :smile: 
It seems this is exactly what I need

Now I'm trying to figure out why I can't create a new middleware
```
quasar new ssrmiddleware mr-middleware
App · ⚠️  Invalid asset type: ssrmiddleware
```
Though I'm using Quasar v2
```
quasar --version
2.2.9
```
But this is another issue )

**Reply by a-salimkhanov** [reply]
Ok, I got it.
I had to upgrade properly :smile: 
[https://next.quasar.dev/start/upgrade-guide](url)

**Reply by ontwikkelfabriek** [reply]
@a-salimkhanov 

Please keep us up to date of your experience with (frontend initiated) server side api request using Quasar's new middleware feature.
